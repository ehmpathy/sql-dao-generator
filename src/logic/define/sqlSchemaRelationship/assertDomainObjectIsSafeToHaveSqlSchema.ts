import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';
import { UserInputError } from '../../UserInputError';

export const assertDomainObjectIsSafeToHaveSqlSchema = ({ domainObject }: { domainObject: DomainObjectMetadata }) => {
  // make sure that entities have "unique" and "updatable" defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_ENTITY) {
    if (!domainObject.decorations.unique?.length)
      throw new UserInputError({
        reason:
          "domain entities must have at least one 'unique' property defined in order for a schema to be generated.",
        domainObjectName: domainObject.name,
        potentialSolution:
          "(note: if its not unique on any natural keys, please specify the 'uuid' as it's only unique property. e.g., `['uuid']`) ",
      });
    if (!Array.isArray(domainObject.decorations.updatable))
      throw new UserInputError({
        reason:
          "domain entities must have their 'updatable' properties defined in order for a schema to be generated. ",
        domainObjectName: domainObject.name,
        potentialSolution:
          '(note: if it has no updatable properties, please specify this with an empty array. e.g., `[]`) ',
      });
  }

  // make sure that value objects do not have unique or updatable defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT) {
    if (domainObject.decorations.unique)
      throw new UserInputError({
        reason:
          "domain value objects must _not_ have their 'unique' properties specified. value objects are unique on all of their properties by definition.",
        domainObjectName: domainObject.name,
      });
    if (domainObject.decorations.updatable)
      throw new UserInputError({
        reason:
          "domain value objects must _not_ have any 'updatable' properties specified. value objects do not have any updatable properties by definition.",
        domainObjectName: domainObject.name,
      });
  }

  // make sure that events have "unique" defined
  if (domainObject.extends === DomainObjectVariant.DOMAIN_EVENT) {
    if (!domainObject.decorations.unique?.length)
      throw new UserInputError({
        reason: "domain events must have at least one 'unique' property defined in order for a schema to be generated",
        domainObjectName: domainObject.name,
      });
  }
};
