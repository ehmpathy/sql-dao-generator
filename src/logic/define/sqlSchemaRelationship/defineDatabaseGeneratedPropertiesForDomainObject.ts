import { DomainObjectMetadata } from 'domain-objects-metadata';
import { isPresent } from 'type-fns';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';

export const defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject = ({
  domainObject,
}: {
  domainObject: DomainObjectMetadata;
}) => {
  const isUniqueOnUuid = domainObject.decorations.unique?.includes('uuid');
  const hasUpdatableProperties =
    (domainObject.decorations.updatable?.length ?? 0) > 0;

  // define which sql schema properties are generated by the db
  return [
    new SqlSchemaPropertyMetadata({
      name: 'id',
      isDatabaseGenerated: true,
      isArray: false,
      isNullable: false,
      isUpdatable: false,
      reference: null,
    }),
    !isUniqueOnUuid
      ? new SqlSchemaPropertyMetadata({
          name: 'uuid',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // "uuid" is only autogenerated if the domain object is not unique on it
    new SqlSchemaPropertyMetadata({
      name: 'created_at',
      isDatabaseGenerated: true,
      isArray: false,
      isNullable: false,
      isUpdatable: false,
      reference: null,
    }),
    hasUpdatableProperties
      ? new SqlSchemaPropertyMetadata({
          name: 'effective_at',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // domain objects only have "effective_at" defined if they're updatable
    hasUpdatableProperties
      ? new SqlSchemaPropertyMetadata({
          name: 'updated_at',
          isDatabaseGenerated: true,
          isArray: false,
          isNullable: false,
          isUpdatable: false,
          reference: null,
        })
      : null, // domain objects only have "updated_at" defined if they're updatable
  ].filter(isPresent);
};
