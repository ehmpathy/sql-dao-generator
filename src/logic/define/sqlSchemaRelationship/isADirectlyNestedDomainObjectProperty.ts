import {
  DomainObjectPropertyMetadata,
  isDomainObjectArrayProperty,
  isDomainObjectReferenceProperty,
} from 'domain-objects-metadata';

import { SqlSchemaPropertyMetadata } from '../../../domain';
import { isAUserDefinedDomainObjectProperty } from './isAUserDefinedDomainObjectProperty';

/**
 * specifies whether this domain object property is a reference to another domain object directly by nesting (in array or solo)
 */
export const isADirectlyNestedDomainObjectProperty = (propertyRelationship: {
  domainObject: DomainObjectPropertyMetadata | null;
  sqlSchema: SqlSchemaPropertyMetadata;
}): propertyRelationship is {
  domainObject: DomainObjectPropertyMetadata;
  sqlSchema: SqlSchemaPropertyMetadata;
} => {
  if (!isAUserDefinedDomainObjectProperty(propertyRelationship)) return false; // if not user defined, then its def not a reference
  if (isDomainObjectReferenceProperty(propertyRelationship.domainObject))
    return true; // if its directly nested, its directly nested
  if (
    isDomainObjectArrayProperty(propertyRelationship.domainObject) &&
    isDomainObjectReferenceProperty(propertyRelationship.domainObject.of)
  )
    return true; // if its an array of directly nested objects, still true
  return false; // otherwise, false
};
