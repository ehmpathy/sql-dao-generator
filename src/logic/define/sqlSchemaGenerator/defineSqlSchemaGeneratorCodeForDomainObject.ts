import { camelCase, noCase } from 'change-case';
import { DomainObjectMetadata, DomainObjectVariant } from 'domain-objects-metadata';
import { isPresent } from 'simple-type-guards';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UserInputError } from '../../UserInputError';
import { isNotADatabaseGeneratedProperty } from '../sqlSchemaRelationship/isNotADatabaseGeneratedProperty';
import { defineSqlSchemaGeneratorCodeForProperty } from './defineSqlSchemaGeneratorCodeForProperty';

export const defineSqlSchemaGeneratorCodeForDomainObject = ({
  domainObject,
  sqlSchemaRelationship,
}: {
  domainObject: DomainObjectMetadata;
  sqlSchemaRelationship: SqlSchemaToDomainObjectRelationship;
}) => {
  // define which sql-schema-generator class we'll be using for defining the schema for this domain object
  const schemaGeneratorClass = (() => {
    if (domainObject.extends === DomainObjectVariant.DOMAIN_ENTITY) return 'Entity';
    if (domainObject.extends === DomainObjectVariant.DOMAIN_VALUE_OBJECT) return 'ValueObject';
    if (domainObject.extends === DomainObjectVariant.DOMAIN_EVENT) return 'Event';
    throw new UserInputError({
      reason: `sql schema can only be created for a domain object which extends DomainEntity, DomainValueObject, or DomainEvent. Found extends '${domainObject.extends}'`,
      domainObjectName: domainObject.name,
    });
  })();

  // define the code for each property
  const schemaGeneratorProperties = sqlSchemaRelationship.properties
    .filter(isNotADatabaseGeneratedProperty) // filter out all db generated properties, since sql-schema-generator will complain if they're included
    .filter((propertyRelationship) => propertyRelationship.domainObject.name !== 'uuid') // we also can't use the "uuid" property, even if its not db generated, since sql-schema-generator will still complain even if entity is unique on it
    .map((propertyRelationship) =>
      defineSqlSchemaGeneratorCodeForProperty({
        domainObject,
        domainObjectProperty: propertyRelationship.domainObject,
        sqlSchemaProperty: propertyRelationship.sqlSchema,
      }),
    );

  // define the code for the unique properties
  const schemaGeneratorUnique =
    sqlSchemaRelationship.decorations.unique.sqlSchema &&
    domainObject.extends !== DomainObjectVariant.DOMAIN_VALUE_OBJECT
      ? `unique: [${sqlSchemaRelationship.decorations.unique.sqlSchema.map((s) => `'${s}'`).join(', ')}],`
      : null;

  // define the imported props from sql-schema-generator based on this
  const schemaGeneratorImports = [schemaGeneratorClass, 'prop']
    .sort((a, b) => (a.toLowerCase() < b.toLowerCase() ? -1 : 1))
    .join(', ');

  // figure out all of the referenced domain objects and define the imports for those
  const referenceImports = sqlSchemaRelationship.properties
    .map((propertyRelationship) => propertyRelationship.sqlSchema.reference)
    .filter(isPresent)
    .map((reference) => `import { ${camelCase(reference.of.name)} } from './${camelCase(reference.of.name)}';`);
  const distinctSortedReferenceImports = [...new Set(referenceImports.sort())];

  // define the code
  const content = `
import { ${schemaGeneratorImports} } from 'sql-schema-generator';
${distinctSortedReferenceImports.length ? ['', ...distinctSortedReferenceImports, ''].join('\n') : ''}
/**
 * sql-schema for the ${noCase(domainObject.extends)} '${domainObject.name}'
 */
export const ${camelCase(domainObject.name)} = new ${schemaGeneratorClass}({
  name: '${sqlSchemaRelationship.name.sqlSchema}',
  properties: {
    ${schemaGeneratorProperties.join('\n    ')}
  },
${schemaGeneratorUnique ? `  ${schemaGeneratorUnique}\n` : ''}});
  `.trim();

  return content;
};
