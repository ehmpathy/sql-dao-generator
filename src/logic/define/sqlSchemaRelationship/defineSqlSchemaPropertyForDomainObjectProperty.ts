import { camelCase, snakeCase } from 'change-case';
import { DomainObjectMetadata, DomainObjectPropertyMetadata, DomainObjectPropertyType } from 'domain-objects-metadata';

import { SqlSchemaPropertyMetadata } from '../../../domain/objects/SqlSchemaPropertyMetadata';
import { SqlSchemaReferenceMethod } from '../../../domain/objects/SqlSchemaReferenceMetadata';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaReferenceForDomainObjectProperty } from './defineSqlSchemaReferenceForDomainObjectProperty';

export const defineSqlSchemaPropertyForDomainObjectProperty = ({
  property,
  domainObject,
  allDomainObjects,
}: {
  property: DomainObjectPropertyMetadata;
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaPropertyMetadata => {
  // sanity check that property name is in camel case
  if (camelCase(property.name) !== property.name)
    throw new UserInputError({
      reason: 'property names must be camel case',
      domainObjectName: domainObject.name,
      domainObjectPropertyName: property.name,
    });

  // sanity check that this is not called for a database generated property (since that should be handled elsewhere)
  const databaseGeneratedProperties = defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({ domainObject });
  if (databaseGeneratedProperties.map(({ name }) => name).includes(snakeCase(property.name)))
    throw new UnexpectedCodePathDetectedError({
      reason: '`defineSqlSchemaPropertyForDomainObjectProperty was called for an db-generated property',
      domainObjectName: domainObject.name,
      domainObjectPropertyName: property.name,
    });

  // define data about this property
  const isUpdatable = domainObject.decorations.updatable?.includes(property.name) ?? false;
  const isNullable = !!property.nullable;
  const isArray = property.type === DomainObjectPropertyType.ARRAY;

  // determine if this is a reference of another domain object
  const reference = defineSqlSchemaReferenceForDomainObjectProperty({
    property,
    domainObject,
    allDomainObjects,
  });

  // define the schema property name
  const schemaPropertyName = (() => {
    const baseSchemaPropertyName = snakeCase(property.name);
    if (reference) {
      if (isArray) {
        if (reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING)
          return `${baseSchemaPropertyName.replace(/s$/, '')}_ids`; // suffix w/ ids, since its an array of fks
        if (reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
          return `${baseSchemaPropertyName.replace(/_uuids$/, '')}_ids`; // suffix w/ ids, since its an array of fks
      }
      if (reference.method === SqlSchemaReferenceMethod.DIRECT_BY_NESTING) return `${baseSchemaPropertyName}_id`; // suffix w/ id, since its a fk
      if (reference.method === SqlSchemaReferenceMethod.IMPLICIT_BY_UUID)
        return `${baseSchemaPropertyName.replace(/_uuid$/, '')}_id`; // suffix w/ id, since its a fk
    }
    return baseSchemaPropertyName; // otherwise, its just the regular name, no modifications
  })();

  return new SqlSchemaPropertyMetadata({
    name: schemaPropertyName,
    isUpdatable,
    isNullable,
    isArray,
    isDatabaseGenerated: false, // this method is only called for properties that are not database generated (as checked above)
    reference,
  });
};
