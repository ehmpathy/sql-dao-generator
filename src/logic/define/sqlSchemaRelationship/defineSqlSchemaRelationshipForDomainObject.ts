import { snakeCase } from 'change-case';
import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { SqlSchemaToDomainObjectRelationship } from '../../../domain/objects/SqlSchemaToDomainObjectRelationship';
import { UnexpectedCodePathDetectedError } from '../../UnexpectedCodePathDetectedError';
import { UserInputError } from '../../UserInputError';
import { assertDomainObjectIsSafeToHaveSqlSchema } from './assertDomainObjectIsSafeToHaveSqlSchema';
import { defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject } from './defineDatabaseGeneratedPropertiesForDomainObject';
import { defineSqlSchemaPropertyForDomainObjectProperty } from './defineSqlSchemaPropertyForDomainObjectProperty';
import { isNotADatabaseGeneratedProperty } from './isNotADatabaseGeneratedProperty';

export const defineSqlSchemaRelationshipForDomainObject = ({
  domainObject,
  allDomainObjects,
}: {
  domainObject: DomainObjectMetadata;
  allDomainObjects: DomainObjectMetadata[];
}): SqlSchemaToDomainObjectRelationship => {
  // figure out some relevant info
  const sqlSchemaName = snakeCase(domainObject.name); // names are in snake case

  // define the properties
  const databaseGeneratedSqlSchemaProperties =
    defineDatabaseGeneratedSqlSchemaPropertiesForDomainObject({
      domainObject,
    });
  const propertiesRelationship = [
    /// autogenerated values
    ...databaseGeneratedSqlSchemaProperties.map((sqlSchemaProperty) => {
      // see if this property was defined on the domain object (i.e., they want the value for it)
      const domainObjectProperty =
        Object.values(domainObject.properties).find(
          (property) => snakeCase(property.name) === sqlSchemaProperty.name,
        ) ?? null;

      // if so, validate that it has the correct type
      if (domainObjectProperty) {
        const expectedType = (() => {
          if (sqlSchemaProperty.name === 'id')
            return DomainObjectPropertyType.NUMBER;
          if (sqlSchemaProperty.name === 'uuid')
            return DomainObjectPropertyType.STRING;
          if (sqlSchemaProperty.name === 'created_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'effective_at')
            return DomainObjectPropertyType.DATE;
          if (sqlSchemaProperty.name === 'updated_at')
            return DomainObjectPropertyType.DATE;
          throw new UnexpectedCodePathDetectedError({
            reason:
              'unexpected database generated property name. this is a bug within sql-schema-generator.', // fail fast if expectations not met
            domainObjectName: domainObject.name,
            domainObjectPropertyName: sqlSchemaProperty.name,
          });
        })();
        if (domainObjectProperty.type !== expectedType)
          throw new UserInputError({
            reason: `reserved database-generated property does not match expected type. expected ${expectedType} but received ${domainObjectProperty.type}`,
            domainObjectName: domainObject.name,
            domainObjectPropertyName: domainObjectProperty.name,
          });
      }

      // return the relationship
      return {
        sqlSchema: sqlSchemaProperty,
        domainObject: domainObjectProperty,
      };
    }),

    // non autogenerated values
    ...Object.values(domainObject.properties)
      .filter(
        (domainObjectProperty) =>
          !databaseGeneratedSqlSchemaProperties
            .map(({ name }) => name)
            .includes(snakeCase(domainObjectProperty.name)), // filter out the db generated values, because we already got them above
      )
      .map((domainObjectProperty) => {
        return {
          sqlSchema: defineSqlSchemaPropertyForDomainObjectProperty({
            property: domainObjectProperty,
            domainObject,
            allDomainObjects,
          }),
          domainObject: domainObjectProperty,
        };
      }),
  ];

  // define the unique keys
  const sqlSchemaUniqueKeys = (() => {
    // define what the unique prop names are, based on what kind of object it is
    const domainObjectPropertyNamesInUniqueKey =
      domainObject.extends === DomainObjectVariant.DOMAIN_LITERAL
        ? propertiesRelationship // if its a literal, the schema is unique on all properties of the object (excluding autogenerated ones like id and uuid)
            .filter(isNotADatabaseGeneratedProperty)
            .map(
              ({ domainObject: domainObjectProperty }) =>
                domainObjectProperty.name,
            )
        : domainObject.decorations.unique; // otherwise, its unique on what was specified

    // convert the domain object prop names into sql schema prop names
    return (
      domainObjectPropertyNamesInUniqueKey?.map((domainObjectPropertyName) => {
        const sqlSchemaPropertyName = propertiesRelationship.find(
          (propertyRelationship) =>
            propertyRelationship.domainObject?.name ===
            domainObjectPropertyName,
        )?.sqlSchema.name;
        if (!sqlSchemaPropertyName)
          throw new UserInputError({
            reason: 'Unique keys must be properties of the domain object.',
            domainObjectName: domainObject.name,
            potentialSolution: ` Was there a typo defining '${
              domainObject.name
            }.unique' with key name '${domainObjectPropertyName}'? (Is it one of ${JSON.stringify(
              Object.keys(domainObject.properties),
            )}?)`,
          });
        return sqlSchemaPropertyName;
      }) ?? null
    );
  })();

  // make sure that this domain object is safe to have a schema defined for it
  assertDomainObjectIsSafeToHaveSqlSchema({ domainObject });

  // return the relationship
  return new SqlSchemaToDomainObjectRelationship({
    name: {
      sqlSchema: sqlSchemaName,
      domainObject: domainObject.name,
    },
    properties: propertiesRelationship,
    decorations: {
      alias: {
        domainObject: domainObject.decorations.alias,
      },
      unique: {
        sqlSchema: sqlSchemaUniqueKeys,
        domainObject: domainObject.decorations.unique,
      },
    },
  });
};
