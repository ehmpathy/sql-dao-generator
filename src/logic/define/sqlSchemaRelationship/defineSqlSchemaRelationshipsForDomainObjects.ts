import { DomainObjectMetadata } from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from './defineSqlSchemaRelationshipForDomainObject';

/**
 * uses the full list of domain objects to define a relationship between its DomainObject representation and its SqlSchema representation
 */
export const defineSqlSchemaRelationshipsForDomainObjects = ({
  domainObjects,
}: {
  domainObjects: DomainObjectMetadata[];
}) =>
  domainObjects.map((domainObject) =>
    defineSqlSchemaRelationshipForDomainObject({
      domainObject,
      allDomainObjects: domainObjects,
    }),
  );
