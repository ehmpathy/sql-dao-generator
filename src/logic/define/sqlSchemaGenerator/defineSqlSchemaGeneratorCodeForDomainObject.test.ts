import {
  DomainObjectMetadata,
  DomainObjectPropertyType,
  DomainObjectVariant,
} from 'domain-objects-metadata';

import { defineSqlSchemaRelationshipForDomainObject } from '../sqlSchemaRelationship/defineSqlSchemaRelationshipForDomainObject';
import { defineSqlSchemaGeneratorCodeForDomainObject } from './defineSqlSchemaGeneratorCodeForDomainObject';

describe('defineSqlSchemaGeneratorCodeForDomainObject', () => {
  describe('imports', () => {
    it('should not have extra new lines if there are no imports', () => {
      // define what we're testing on
      const domainObject = new DomainObjectMetadata({
        name: 'Geocode',
        extends: DomainObjectVariant.DOMAIN_LITERAL,
        properties: {
          latitude: {
            name: 'latitude',
            type: DomainObjectPropertyType.NUMBER,
          },
          longitude: {
            name: 'longitude',
            type: DomainObjectPropertyType.NUMBER,
          },
        },
        decorations: {
          unique: null,
          updatable: null,
        },
      });
      const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
        domainObject,
        allDomainObjects: [domainObject],
      });

      // run it
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        domainObject,
        sqlSchemaRelationship,
      });

      // check import lines
      const importLines = code
        .split('\n')
        .filter((line) => line.includes('import {'));
      expect(importLines.length).toEqual(1); // should only have one import
      expect(importLines[0]).toContain("from 'sql-schema-generator';"); // and it should be the one from the sql-schema-generator

      // check that there's not an extra newline after the import
      const lines = code.split('\n');
      expect(lines[0]).toContain('import {'); // should be the first import
      expect(lines[1]).toEqual(''); // should be a newline
      expect(lines[2]).not.toEqual(''); // should not be a newline

      // take a snapshot for documenting example
      expect(code).toMatchSnapshot();
    });
    it('should create an import for each directly nested and uuid inferred reference, sorted', () => {
      // define what we're testing on
      const domainObject = new DomainObjectMetadata({
        name: 'Train',
        extends: DomainObjectVariant.DOMAIN_ENTITY,
        properties: {
          id: {
            name: 'id',
            type: DomainObjectPropertyType.NUMBER,
            required: false,
          },
          uuid: {
            name: 'uuid',
            type: DomainObjectPropertyType.STRING,
            required: false,
          },
          homeStationGeocode: {
            name: 'homeStation',
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Station',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
          },
          badges: {
            name: 'badges',
            type: DomainObjectPropertyType.ARRAY,
            of: {
              type: DomainObjectPropertyType.REFERENCE,
              of: {
                name: 'Badge',
                extends: DomainObjectVariant.DOMAIN_LITERAL,
              },
            },
          },
          locomotiveUuids: {
            name: 'locomotiveUuids',
            type: DomainObjectPropertyType.ARRAY,
            of: {
              type: DomainObjectPropertyType.STRING,
            },
          },
          leadEngineerUuid: {
            name: 'leadEngineerUuid',
            type: DomainObjectPropertyType.STRING,
          },
        },
        decorations: {
          unique: ['uuid'],
          updatable: [
            'homeStation',
            'badges',
            'locomotiveUuids',
            'leadEngineerUuid',
          ],
        },
      });
      const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
        domainObject,
        allDomainObjects: [
          domainObject,
          { name: 'Locomotive', extends: DomainObjectVariant.DOMAIN_ENTITY },
          { name: 'Engineer', extends: DomainObjectVariant.DOMAIN_ENTITY },
        ] as DomainObjectMetadata[],
      });

      // run it
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        domainObject,
        sqlSchemaRelationship,
      });

      // check import lines
      const importLines = code
        .split('\n')
        .filter((line) => line.includes('import {'));
      expect(importLines.length).toEqual(5); // should have 5 import lines; one for schema generator, 4 for domain entity references
      expect(importLines[0]).toContain("from 'sql-schema-generator';"); // first should be schema generator
      expect(importLines[1]).toContain("from './badge'"); // b, is first in alphabetical order of subsequent imports
      expect(importLines[2]).toContain("from './engineer'"); // e, is next
      expect(importLines[3]).toContain("from './locomotive'"); // l, is after that
      expect(importLines[4]).toContain("from './station'"); // s, is last

      // check that there's a new line after schema-generator import, no newlines between reference imports, and a newline after reference imports
      const lines = code.split('\n');
      expect(lines[0]).toContain('import {'); // should be the first import
      expect(lines[1]).toEqual(''); // should be a newline
      expect(lines[2]).toContain('import {'); // should be the an import
      expect(lines[3]).toContain('import {'); // should be the an import
      expect(lines[4]).toContain('import {'); // should be the an import
      expect(lines[5]).toContain('import {'); // should be the an import
      expect(lines[6]).toEqual(''); // should be a newline
      expect(lines[7]).not.toEqual(''); // should not be a newline

      // take a snapshot for documenting example
      expect(code).toMatchSnapshot();
    });
  });
  describe('domain object variants', () => {
    it('should create a correct looking sql-schema-generator object for a domain-literal', () => {
      // define what we're testing on
      const domainObject = new DomainObjectMetadata({
        name: 'Geocode',
        extends: DomainObjectVariant.DOMAIN_LITERAL,
        properties: {
          id: {
            name: 'id',
            type: DomainObjectPropertyType.NUMBER,
            required: false,
          },
          uuid: {
            name: 'uuid',
            type: DomainObjectPropertyType.STRING,
            required: false,
          },
          latitude: { name: 'latitude', type: DomainObjectPropertyType.NUMBER },
          longitude: {
            name: 'longitude',
            type: DomainObjectPropertyType.NUMBER,
          },
        },
        decorations: { unique: null, updatable: null },
      });
      const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
        domainObject,
        allDomainObjects: [domainObject],
      });

      // run it
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        domainObject,
        sqlSchemaRelationship,
      });
      expect(code).toContain(
        "import { Literal, prop } from 'sql-schema-generator'",
      );
      expect(code).toContain('export const geocode: Literal = new Literal({');
      expect(code).toContain("name: 'geocode'");
      expect(code).not.toContain('id:');
      expect(code).not.toContain('uuid:');
      expect(code).toContain('latitude: prop.NUMERIC()');
      expect(code).toContain('longitude: prop.NUMERIC()');
      expect(code).not.toContain('unique:');
      expect(code).toMatchSnapshot(); // log example
    });

    it('should create a correct looking sql-schema-generator object for a domain-entity', () => {
      // define what we're testing on
      const domainObject = new DomainObjectMetadata({
        name: 'Carriage',
        extends: DomainObjectVariant.DOMAIN_ENTITY,
        properties: {
          id: {
            name: 'id',
            type: DomainObjectPropertyType.NUMBER,
            required: false,
          },
          uuid: {
            name: 'uuid',
            type: DomainObjectPropertyType.STRING,
            required: false,
          },
          cin: {
            name: 'cin',
            type: DomainObjectPropertyType.STRING,
            required: true,
          },
          carries: {
            name: 'carries',
            type: DomainObjectPropertyType.ENUM,
            of: ['PASSENGER', 'FREIGHT'],
            required: true,
          },
          capacity: {
            name: 'capacity',
            type: DomainObjectPropertyType.NUMBER,
            nullable: true,
          },
        },
        decorations: {
          unique: ['cin'],
          updatable: [],
        },
      });
      const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
        domainObject,
        allDomainObjects: [domainObject],
      });

      // run it
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        domainObject,
        sqlSchemaRelationship,
      });
      expect(code).toContain(
        "import { Entity, prop } from 'sql-schema-generator'",
      );
      expect(code).toContain('export const carriage: Entity = new Entity({');
      expect(code).toContain("name: 'carriage'");
      expect(code).not.toContain('id:'); // should be filtered out
      expect(code).not.toContain('uuid:'); // should be filtered out
      expect(code).toContain('cin: prop.VARCHAR(),');
      expect(code).toContain("carries: prop.ENUM(['PASSENGER', 'FREIGHT']),");
      expect(code).toContain(
        'capacity: { ...prop.NUMERIC(), nullable: true },',
      );
      expect(code).toContain("unique: ['cin'],");
      expect(code).toMatchSnapshot();
    });
    it('should create a correct looking sql-schema-generator object for a domain-event', () => {
      // define what we're testing on
      const domainObject = new DomainObjectMetadata({
        name: 'TrainLocatedEvent',
        extends: DomainObjectVariant.DOMAIN_EVENT,
        properties: {
          id: {
            name: 'id',
            type: DomainObjectPropertyType.NUMBER,
            required: false,
          },
          trainUuid: {
            name: 'trainUuid',
            type: DomainObjectPropertyType.STRING,
            required: true,
          },
          occurredAt: {
            name: 'occurredAt',
            type: DomainObjectPropertyType.DATE,
            required: true,
          },
          geocode: {
            name: 'geocode',
            type: DomainObjectPropertyType.REFERENCE,
            of: {
              name: 'Geocode',
              extends: DomainObjectVariant.DOMAIN_LITERAL,
            },
            required: true,
          },
        },
        decorations: {
          unique: ['trainUuid', 'occurredAt'],
          updatable: [],
        },
      });
      const sqlSchemaRelationship = defineSqlSchemaRelationshipForDomainObject({
        domainObject,
        allDomainObjects: [
          domainObject,
          { name: 'Train', extends: DomainObjectVariant.DOMAIN_ENTITY },
        ] as DomainObjectMetadata[],
      });

      // run it
      const code = defineSqlSchemaGeneratorCodeForDomainObject({
        domainObject,
        sqlSchemaRelationship,
      });
      expect(code).toContain(
        "import { Event, prop } from 'sql-schema-generator'",
      );
      expect(code).toContain(
        'export const trainLocatedEvent: Event = new Event({',
      );
      expect(code).toContain("name: 'train_located_event'");
      expect(code).not.toContain(' id:'); // should be filtered out
      expect(code).not.toContain(' uuid:'); // should be filtered out
      expect(code).toContain('train_id: prop.REFERENCES(train),');
      expect(code).toContain('occurred_at: prop.TIMESTAMPTZ(),');
      expect(code).toContain('geocode_id: prop.REFERENCES(geocode),');
      expect(code).toContain("unique: ['train_id', 'occurred_at'],");
      expect(code).toMatchSnapshot();
    });
  });
});
