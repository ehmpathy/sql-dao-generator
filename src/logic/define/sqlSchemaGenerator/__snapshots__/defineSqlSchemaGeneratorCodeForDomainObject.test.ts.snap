// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`defineSqlSchemaGeneratorCodeForDomainObject domain object variants should create a correct looking sql-schema-generator object for a domain-entity 1`] = `
"import { Entity, prop } from 'sql-schema-generator';

/**
 * sql-schema for the domain entity 'Carriage'
 */
export const carriage: Entity = new Entity({
  name: 'carriage',
  properties: {
    cin: prop.VARCHAR(),
    carries: prop.ENUM(['PASSENGER', 'FREIGHT']),
    capacity: { ...prop.NUMERIC(), nullable: true },
  },
  unique: ['cin'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject domain object variants should create a correct looking sql-schema-generator object for a domain-event 1`] = `
"import { Event, prop } from 'sql-schema-generator';

import { geocode } from './geocode';
import { train } from './train';

/**
 * sql-schema for the domain event 'TrainLocatedEvent'
 */
export const trainLocatedEvent: Event = new Event({
  name: 'train_located_event',
  properties: {
    train_id: prop.REFERENCES(train),
    occurred_at: prop.TIMESTAMPTZ(),
    geocode_id: prop.REFERENCES(geocode),
  },
  unique: ['train_id', 'occurred_at'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject domain object variants should create a correct looking sql-schema-generator object for a domain-value-object 1`] = `
"import { prop, ValueObject } from 'sql-schema-generator';

/**
 * sql-schema for the domain value object 'Geocode'
 */
export const geocode: ValueObject = new ValueObject({
  name: 'geocode',
  properties: {
    latitude: prop.NUMERIC(),
    longitude: prop.NUMERIC(),
  },
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject imports should create an import for each directly nested and uuid inferred reference, sorted 1`] = `
"import { Entity, prop } from 'sql-schema-generator';

import { badge } from './badge';
import { engineer } from './engineer';
import { locomotive } from './locomotive';
import { station } from './station';

/**
 * sql-schema for the domain entity 'Train'
 */
export const train: Entity = new Entity({
  name: 'train',
  properties: {
    home_station_id: { ...prop.REFERENCES(station), updatable: true },
    badge_ids: { ...prop.ARRAY_OF(prop.REFERENCES(badge)), updatable: true },
    locomotive_ids: { ...prop.ARRAY_OF(prop.REFERENCES(locomotive)), updatable: true },
    lead_engineer_id: { ...prop.REFERENCES(engineer), updatable: true },
  },
  unique: ['uuid'],
});"
`;

exports[`defineSqlSchemaGeneratorCodeForDomainObject imports should not have extra new lines if there are no imports 1`] = `
"import { prop, ValueObject } from 'sql-schema-generator';

/**
 * sql-schema for the domain value object 'Geocode'
 */
export const geocode: ValueObject = new ValueObject({
  name: 'geocode',
  properties: {
    latitude: prop.NUMERIC(),
    longitude: prop.NUMERIC(),
  },
});"
`;
